pkg/config/config.go

```diff
@@ -0,0 +1,327 @@
+			Default:  def,
+		}),
+	})
+
+}
+
+func addPeerGroups(ctx context.Context, bgpServer *server.BgpServer, addedPg []config.PeerGroup) {
+	for _, pg := range addedPg {
+		log.Infof("PeerGroup %s is added", pg.Config.PeerGroupName)
+		if err := bgpServer.AddPeerGroup(ctx, &api.AddPeerGroupRequest{
+			PeerGroup: config.NewPeerGroupFromConfigStruct(&pg),
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func deletePeerGroups(ctx context.Context, bgpServer *server.BgpServer, deletedPg []config.PeerGroup) {
+	for _, pg := range deletedPg {
+		log.Infof("PeerGroup %s is deleted", pg.Config.PeerGroupName)
+		if err := bgpServer.DeletePeerGroup(ctx, &api.DeletePeerGroupRequest{
+			Name: pg.Config.PeerGroupName,
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func updatePeerGroups(ctx context.Context, bgpServer *server.BgpServer, updatedPg []config.PeerGroup) bool {
+	for _, pg := range updatedPg {
+		log.Infof("PeerGroup %s is updated", pg.Config.PeerGroupName)
+		if u, err := bgpServer.UpdatePeerGroup(ctx, &api.UpdatePeerGroupRequest{
+			PeerGroup: config.NewPeerGroupFromConfigStruct(&pg),
+		}); err != nil {
+			log.Warn(err)
+		} else {
+			return u.NeedsSoftResetIn
+		}
+	}
+	return false
+}
+
+func addDynamicNeighbors(ctx context.Context, bgpServer *server.BgpServer, dynamicNeighbors []config.DynamicNeighbor) {
+	for _, dn := range dynamicNeighbors {
+		log.Infof("Dynamic Neighbor %s is added to PeerGroup %s", dn.Config.Prefix, dn.Config.PeerGroup)
+		if err := bgpServer.AddDynamicNeighbor(ctx, &api.AddDynamicNeighborRequest{
+			DynamicNeighbor: &api.DynamicNeighbor{
+				Prefix:    dn.Config.Prefix,
+				PeerGroup: dn.Config.PeerGroup,
+			},
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func addNeighbors(ctx context.Context, bgpServer *server.BgpServer, added []config.Neighbor) {
+	for _, p := range added {
+		log.Infof("Peer %v is added", p.State.NeighborAddress)
+		if err := bgpServer.AddPeer(ctx, &api.AddPeerRequest{
+			Peer: config.NewPeerFromConfigStruct(&p),
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func deleteNeighbors(ctx context.Context, bgpServer *server.BgpServer, deleted []config.Neighbor) {
+	for _, p := range deleted {
+		log.Infof("Peer %v is deleted", p.State.NeighborAddress)
+		if err := bgpServer.DeletePeer(ctx, &api.DeletePeerRequest{
+			Address: p.State.NeighborAddress,
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func updateNeighbors(ctx context.Context, bgpServer *server.BgpServer, updated []config.Neighbor) bool {
+	for _, p := range updated {
+		log.Infof("Peer %v is updated", p.State.NeighborAddress)
+		if u, err := bgpServer.UpdatePeer(ctx, &api.UpdatePeerRequest{
+			Peer: config.NewPeerFromConfigStruct(&p),
+		}); err != nil {
+			log.Warn(err)
+		} else {
+			return u.NeedsSoftResetIn
+		}
+	}
+	return false
+}
+
+func ApplyInitialConfig(ctx context.Context, bgpServer *server.BgpServer, newConfig *config.BgpConfigSet, isGracefulRestart bool) *config.BgpConfigSet {
```

Comment:
The one thing I might suggest is to add `error` to the list of return values to this function and to `UpdateConfig`. Then, error handling could be modified in the future without having to change the API.